viewing_party/party.py

```diff
@@ -1,23 +1,193 @@
+    # watch list, for example, because it was always already set up. Thoughts appreciated.
+
+    user_watched = set()
+
+    for movie in user_data['watched']:
+        user_watched.add((movie['title'], movie['genre']))
+
+    return user_watched
+
+def create_friends_watched_set(user_data):
+    friends_watched = set()
+
+    for friend in user_data['friends']:
+        for movie in friend['watched']:
+            friends_watched.add((movie['title'], movie['genre']))
+    
+    return friends_watched
+
+def get_unique_watched(user_data):
+    watched = user_data['watched']
+    unique_movies = []
+
+    user_unique_watched = (create_user_watched_set(user_data)
+        - create_friends_watched_set(user_data))
+
+    for movie in watched:
+        for entry in user_unique_watched:
+            if entry == (movie['title'], movie['genre']):
+                unique_movies.append(movie)
+
+    return unique_movies
+
+def get_friends_unique_watched(user_data):
+    friends = user_data['friends']
+    friends_unique_movies = []
+
+    friends_unique_watched = (create_friends_watched_set(user_data) - 
+        create_user_watched_set(user_data))
+    
+    for friend in friends:
+        for movie in friend['watched']:
+            for entry in friends_unique_watched:
+                if entry == (movie['title'], movie['genre']):
+                    if movie in friends_unique_movies:
+                        continue
+                    else:
+                        friends_unique_movies.append(movie)
+
+    return friends_unique_movies
+
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    
+    # making sure that this is robust to subscriptions not existing because
+    # it's not in the Arrange section of the test where Sonya has no friends.
+    
+    subscriptions = user_data.get('subscriptions', [])
+    available_recs = []
+
+    friends_unique = get_friends_unique_watched(user_data)
+
+    for movie in friends_unique:
+        if movie['host'] in subscriptions:
+            available_recs.append(movie)
+
+    return available_recs
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    rec_genre = get_most_watched_genre(user_data)
+    available_recs = get_available_recs(user_data)
+    movie_recs_by_genre = []
+
+    for movie in available_recs:
+        if movie['genre'] == rec_genre:
+            movie_recs_by_genre.append(movie)
+    
+    return movie_recs_by_genre
+
+def get_rec_from_favorites(user_data):
+    available_recs = get_unique_watched(user_data)
+    recs_in_favorites = []
+
+    for movie in available_recs:
+        if movie in user_data['favorites']:
+            recs_in_favorites.append(movie)

```

Comment:
Another opportunity for a list comprehension here, as well as potentially making a set out of the favorites using your tuple layout (especially if you make a helper for building the tuple).